common/nym_offline_compact_ecash/src/scheme/setup.rs

```diff
@@ -0,0 +1,560 @@
+use bls12_381::{G1Affine, G1Projective, G2Affine, G2Prepared, G2Projective, Scalar};
+use ff::Field;
+use group::{Curve, GroupEncoding};
+use rand::thread_rng;
+
+use crate::error::{CompactEcashError, Result};
+use crate::{constants, Base58};
+
+use crate::scheme::keygen::{SecretKeyAuth, VerificationKeyAuth};
+use crate::traits::Bytable;
+use crate::utils::{check_bilinear_pairing, generate_lagrangian_coefficients_at_origin};
+use crate::utils::{hash_g1, try_deserialize_g1_projective};
+use itertools::Itertools;
+use rayon::prelude::*;
+
+#[derive(Debug)]
+pub struct GroupParameters {
+    /// Generator of the G1 group
+    g1: G1Affine,
+    /// Generator of the G2 group
+    g2: G2Affine,
+    /// Additional generators of the G1 group
+    gammas: Vec<G1Projective>,
+    // Additional generator of the G1 group
+    delta: G1Projective,
+    /// Precomputed G2 generator used for the miller loop
+    _g2_prepared_miller: G2Prepared,
+}
+
+impl GroupParameters {
+    pub fn new() -> GroupParameters {
```

Comment:
my take: global constants are ok but global mutable state not. This looks like its always initid to hardcoded values so it probably falls under the first category?

Reply:
If the scope of the library is only our case of online ecash, then that `GroupParameters` is gonna be fixed. Most of the library is coded that way, so I have no objection for doing that. 
I'm gonna keep that for later, to avoid messing up the PR stack by propagating those changes

Reply:
Fixed in ebd882992abaa333d3f6a3d61411526ac4830a14